pyomo/contrib/appsi/solvers/maingo.py

```diff
@@ -0,0 +1,482 @@
+from collections import namedtuple
+import logging
+import math
+import sys
+from typing import Optional, List, Dict
+
+from pyomo.contrib.appsi.base import (
+    PersistentSolver,
+    Results,
+    TerminationCondition,
+    MIPSolverConfig,
+    PersistentBase,
+    PersistentSolutionLoader,
+)
+from pyomo.contrib.appsi.cmodel import cmodel, cmodel_available
+from pyomo.common.collections import ComponentMap
+from pyomo.common.config import ConfigValue, NonNegativeInt
+from pyomo.common.dependencies import attempt_import
+from pyomo.common.errors import PyomoException
+from pyomo.common.log import LogStream
+from pyomo.common.tee import capture_output, TeeStream
+from pyomo.common.timing import HierarchicalTimer
+from pyomo.core.base import SymbolMap, NumericLabeler, TextLabeler
+from pyomo.core.base.constraint import _GeneralConstraintData
+from pyomo.core.base.expression import ScalarExpression
+from pyomo.core.base.param import _ParamData
+from pyomo.core.base.sos import _SOSConstraintData
+from pyomo.core.base.var import Var, ScalarVar, _GeneralVarData
+import pyomo.core.expr.expr_common as common
+import pyomo.core.expr as EXPR
+from pyomo.core.expr.numvalue import (
+    value,
+    is_constant,
+    is_fixed,
+    native_numeric_types,
+    native_types,
+    nonpyomo_leaf_types,
+)
+from pyomo.core.kernel.objective import minimize, maximize
+from pyomo.core.staleflag import StaleFlagManager
+from pyomo.repn.util import valid_expr_ctypes_minlp
+
+
+def _import_SolverModel():
+    try:
+        from . import maingo_solvermodel
+    except ImportError:
+        raise
+    return maingo_solvermodel
+
+
+maingo_solvermodel, solvermodel_available = attempt_import(
+    "maingo_solvermodel", importer=_import_SolverModel
+)
+
+MaingoVar = namedtuple("MaingoVar", "type name lb ub init")
+
+logger = logging.getLogger(__name__)
+
+
+def _import_maingopy():
+    try:
+        import maingopy
+    except ImportError:
+        MAiNGO._available = MAiNGO.Availability.NotFound
+        raise
+    return maingopy
+
+
+maingopy, maingopy_available = attempt_import("maingopy", importer=_import_maingopy)
+
+
+class MAiNGOConfig(MIPSolverConfig):
+    def __init__(
+        self,
+        description=None,
+        doc=None,
+        implicit=False,
+        implicit_domain=None,
+        visibility=0,
+    ):
+        super(MAiNGOConfig, self).__init__(
+            description=description,
+            doc=doc,
+            implicit=implicit,
+            implicit_domain=implicit_domain,
+            visibility=visibility,
+        )
+
+        self.declare("logfile", ConfigValue(domain=str))
```

Comment:
You can specify a default here and remove L#94:
```python
self.declare("logfile", ConfigValue(domain=str, default=None)
```

Reply:
Thank you for the tip, done.